### ISSUE TITLE
[RyuJIT/ARM32] Support fast tail call optimization

### ISSUE THREAD

**Issue by hseok-oh** [dotnet/coreclr#13828]
ARM32 ruyjit backend doesn't support fast tail call optimization: `FEATURE_FASTTAILCALL`.
To introduce this feature on RyuJIT/ARM32, we need additional implementation on morphing(fasttailcall decision), lowering, and codegen phase.

**Comment by hseok-oh** [comment]
cc/ @dotnet/arm32-contrib

**Comment by BruceForstall** [comment]
@hseok-oh What else is left to implement (if anything)?

**Comment by hseok-oh** [comment]
@BruceForstall Remain implementation to enable fast tail call is `fgCanFastTailCall` on morphing phase check callee can optimize using fast tail call. I'll update PR #14056 for this.

**Comment by hseok-oh** [comment]
In PR #14056, 

- Not use fast tail call when callee use floating point register argument (difficult to calculate stack size)
- Not use fast tail call when callee use split struct argument 
- Not use fast tail call when callee has struct arguments and stack arguments

**Comment by lt72** [comment]
@sywhang FYI
@BruceForstall: do you know when you will get a chance to work on this issue?

**Comment by BruceForstall** [comment]
I plan to review and (hopefully) merge the outstanding PR within the next couple days. I had been holding off until RyuJIT/arm32 became the default, which happened yesterday.

**Comment by lt72** [comment]
@BruceForstall: was this completed already?

**Comment by BruceForstall** [comment]
@lt72 No. If you look at the PR to enable it, https://github.com/dotnet/coreclr/pull/14056, there are lots of test failures with the current change that need to be addressed.

**Comment by BruceForstall** [comment]
Waiting until https://github.com/dotnet/coreclr/pull/16039 is resolved before reviewing this again.

**Comment by alpencolt** [comment]
There is remaning issues as I understand.
@BruceForstall could you assign this issue to me? I will take a look little bit later.

**Comment by lt72** [comment]
@alpencolt: any progress with this one?
@BruceForstall FYI

**Comment by alpencolt** [comment]
As I see this issue is in 2.1 milestone.
@BruceForstall @lt72 is there any deadlines for this feature? Unfortunately I don't have enough time for resolving all crashes and implementing missed functionality at the current moment.
Is it OK to finish all works on tail calls in April-May? If it's late I will ask other person to continue work.

**Comment by BruceForstall** [comment]
@alpencolt There is roughly a month left to finish all development work for 2.1. We are focused on increasing our test coverage and fixing bugs. Since this is purely an optimization, I don't view it as essential to 2.1, so it is likely to be pushed out to Future. In fact, it might be considered too risky to enable it even now, given that it is getting late in the cycle.

Note that there was confusion here regarding a dependency with profiling: there should be no such dependency. Profiling should work with the "non-fast tailcall" path we should be generating now (and if not, we should fix it anyway). When we do finalize this fast tailcall feature, we will also need to make sure it works with profiling.

Given all this, I will go ahead and move it to Future now.

**Comment by BruceForstall** [comment]
@jashook @jakobbotsch Is enabling fast tailcall for arm32 in scope for your work?

**Comment by jakobbotsch** [comment]
It is not planned, no.

**Comment by jashook** [comment]
x86 windows is also out of scope.
